reply.lua

```diff
@@ -0,0 +1,166 @@
+--
+-- SPDX-FileCopyrightText: (c) 2024 Ring Zero Desenvolvimento de Software LTDA
+-- SPDX-License-Identifier: GPL-2.0-only
+--
+
+local data   = require("data")
+local linux  = require("linux")
+local socket = require("socket")
+local pack   = require("dome/pack")
+local config = require("dome/config")
+
+local PACKET = socket.af.PACKET
+local RAW    = socket.sock.RAW
+
+local ETH_HLEN  = 14
+local ETH_P_ALL = 0x0003
+
+local IPPROTO_TCP = 0x06
+
+local RST = 1 << 2
+local ACK = 1 << 4
+
+local packer   = pack.packer
+local unpacker = pack.unpacker
+
+local reply = {}
+
+local ntoh16 = linux.ntoh16
+local function checksum(data, base, length)
+	local base = base or 0
+	local length = length or #data - base
+	local sum = 0
+
+	for offset = 0, length - 1, 2 do
+		sum = sum + ntoh16(data:getuint16(base + offset))
+	end
+	if length % 2 == 1 then
+		sum = sum + data:getu8(base + length - 1)
+	end
+	sum = sum & 0xffffffff
+	while sum >> 16 ~= 0 do
+		sum = (sum >> 16) + sum & 0xffff
+	end
+
+	return ~sum & 0xffff
+end
+
+local function ethhdr(packet)
+	local getstr, _, getbe16 = unpacker(packet)
+	return {
+		dst     = getstr(0, 6),
+		src     = getstr(6, 6),
+		ethtype = getbe16(12),
+	}
+end
+
+local function iphdr(packet)
+	local getstr, getu8, getbe16, getbe32 = unpacker(packet, ETH_HLEN)
+	return {
+		version  = getu8(0) >> 4,
+		ihl      = getu8(0) & 0xf,
+		tos      = getu8(1),
+		tot_len  = getbe16(2),
+		id       = getbe16(4),
+		frag_off = getbe16(6),
+		ttl      = getu8(8),
+		protocol = getu8(9),
+		check    = getbe16(10),
+		saddr    = getbe32(12),
+		daddr    = getbe32(16),
+	}
+end
+
+local function tcphdr(packet)
+	local ihl = packet:getbyte(ETH_HLEN) & 0xf
+	local base = ETH_HLEN + ihl * 4
+	local _, getu8, getbe16, getbe32 = unpacker(packet, base)
+	return {
+		source  = getbe16(0),
+		dest    = getbe16(2),
+		seq     = getbe32(4),
+		ack_seq = getbe32(8),
+		doff    = getu8(12) >> 4,
+		flags   = getu8(13),
+		window  = getbe16(14),
+		check   = getbe16(16),
+	}, base
+end
+
+local function pseudohdr(ip, tcp)
+	local pseudo = data.new(12 + tcp.doff * 4)
+	local _, setu8, setbe16, setbe32 = packer(pseudo)
+
+	setbe32(0, ip.saddr)
+	setbe32(4, ip.daddr)
+	setu8(8, 0)
+	setu8(9, IPPROTO_TCP)
+	setbe16(10, tcp.doff * 4)
+
+	return pseudo
+end
+
+local function edit_eth(packet)
+	local setstr, _, setbe16 = packer(packet, 0)
+	local eth = ethhdr(packet)
+
+	setstr(0, eth.src)
+	setstr(6, eth.dst)
+	setbe16(12, eth.ethtype)
+end
+
+local function edit_ip(packet, tcp_len)
+	local _, setu8, setbe16, setbe32 = packer(packet, ETH_HLEN)
+	local ip = iphdr(packet, ETH_HLEN)
+	local size = ip.ihl * 4
+
+	setbe16(2, size + tcp_len) -- tot_len = 40
+	setbe16(4, 0)              -- id = 0
+	setbe16(6, 0)              -- frag_off = 0
+	setbe32(12, ip.daddr)      -- saddr = daddr
+	setbe32(16, ip.saddr)      -- daddr = saddr
+	setbe16(10, 0)             -- checksum = 0
+	setbe16(10, checksum(packet, ETH_HLEN, ip.ihl * 4))
```

Comment:
Isn't `ip.ihl * 4` already stored in `size`?

Reply:
oops